### DISCUSSION TITLE
Lilygo T-Beam vr.1.1 Output power issue

### DISCUSSION THREAD

**Discussion by micfuc** [jgromes/RadioLib#1078] [General]
Hi, I am trying to check the output power for the lilygo T-Beam vr. 1.1 device... it is using an sx1268 like LoRa chip and it works apparently OK either using the RadioLib  library and using the old classical LoRa library (https://github.com/sandeepmistry/arduino-LoRa) tipically used by most of the old SW made to run on this device. 

What I discovered is that the maximun output power I can achive from the device with the old LoRa library on two samples of the device I have in my hands is around 19.2 dbm, while using the RadioLib it is maximun 15.8 dbm...  

Making a long story short, I did a very simple program to set the LoRa chip to transmit a long LoRa packet with both libraries and  this confirmed exactly this fact... 
Then I made the same test with the same program on a custom ESP32 device I made using  an RFM-98 and an SX1268 module from different makers and confirmed exactly the same fact: radiolib apparently does not allow to use the full  20dbm power capability of the sx1278 semtech chip... 
I made also a check of the actual code  implementing the setOutputPower  in the two libraries and apparently they do the same operations... also tryed to set up to the max setCurrentLimit without any change...

Do you have any suggestion where to investigate to fix this issue ?
I add my test code for reference.

Thanks in advance for any help
Michele I8FUC

========================================================================================

```c++
#include <SPI.h>
#include <Arduino.h>

//#define RADIOLIB

#ifndef RADIOLIB
#include <LoRa.h>
#else
#include <RadioLib.h>
#endif

int counter = 0;

#define LORA_SCK 5      // GPIO5  - SX1278 SCK
#define LORA_MISO 19    // GPIO19 - SX1278 MISO
#define LORA_MOSI 27    // GPIO27 - SX1278 MOSI
#define LORA_CS 18      // GPIO18 - SX1278 CS
#define LORA_RST 23     // GPIO23 - SX1278 RST
#define LORA_IRQ 26     // GPIO26 - SX1278 IO0

#ifdef RADIOLIB
  SX1278 LoRa = new Module(LORA_CS, 0 , LORA_RST,0);
#endif


void setup() {
  Serial.begin(115200);
  while (!Serial);
#ifdef RADIOLIB
  Serial.println("LoRa Sender - RadioLib");
  SPI.begin(LORA_SCK, LORA_MISO, LORA_MOSI, LORA_CS);

  LoRa.setCurrentLimit(230.0);
  LoRa.setOutputPower(20);

  int state = LoRa.begin(433.775, 125.0 , 12 , 5 , 0x12, 20 , 20, 0 );
  if (state == RADIOLIB_ERR_NONE) {
    int LoRaChipVersion = LoRa.getChipVersion();
    Serial.printf("success (LoRaChipVersion=%d)!\r", LoRaChipVersion);
    }
  else {
    Serial.printf("failed, code %d: Aborting... please fix and reboot 1 !\r ",state );
    };

  LoRa.setSyncWord(0x12);
  LoRa.setSpreadingFactor(12);
  LoRa.setBandwidth(125.00);
  LoRa.setCodingRate(5);
  LoRa.setOutputPower(20);
  LoRa.setPreambleLength(20);
  LoRa.setGain(0);
  //  if(sf > 10) lora_1g.forceLDRO(true);
  //  LoRa.forceLDRO(false);
  //  LoRa.autoLDRO() ;
  LoRa.setCRC(true);

#else
  Serial.println("LoRa Sender - LoRa_Lib");
  LoRa.setPins(LORA_CS, LORA_RST, LORA_IRQ);

  long freq = 433775000;
  if (!LoRa.begin(freq)) {
     Serial.println("Starting LoRa failed!");
  }
  
  LoRa.setSpreadingFactor(12);
  LoRa.setSignalBandwidth(125000);
  LoRa.setCodingRate4(5);
  LoRa.enableCrc();

  LoRa.setTxPower(20);
#endif

}


#ifndef RADIOLIB
void loop() {
  Serial.print("Sending packet: ");
  Serial.println(counter);

  // send packet
  LoRa.beginPacket();
  LoRa.print("Hello World!##################################################################################");
  LoRa.print(counter);
  LoRa.endPacket();
  counter++;
  delay(100);
}

#else
void loop() {
  Serial.print(F("[SX1278] Transmitting packet ... "));
  // you can transmit C-string or Arduino string up to
  // 255 characters long
  String str = "Hello World! #" + String(counter++);
 // int state = LoRaWANBands.transmit(("Hello World! #").c_str());
  int state = LoRa.transmit("Hello World! ################################################################################");

  if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F(" success!"));
    } else if (state == RADIOLIB_ERR_PACKET_TOO_LONG) {
    // the supplied packet was longer than 256 bytes
    Serial.println(F("too long!"));
    } 
  else if (state == RADIOLIB_ERR_TX_TIMEOUT) {
    // timeout occurred while transmitting packet
    Serial.println(F("timeout!"));
    } 
  else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);
    }

  // wait for a second before transmitting again
  delay(10);
}

#endif
```

**Comment by StevenCellist** [comment]
Not my ballpark, but just out of interest: if you set both to e.g. 15dBm, do they result in equal output, or does that also differ substantially between both libs?

**Reply by micfuc** [reply]
Hi,
Just tested with 15dbm setup:
old LoRa : 16.3 dbm
Radiolib: 15.8 dbm

so they agree...  the problem is apparently only for the 20dbm that requires for the SX1268 specifications a special setup for the PA stage in the semtech chip...
73

**Comment by jgromes** [comment]
>  it is using an sx1268 like LoRa chip and it works apparently OK either using the RadioLib library and using the old classical LoRa library 

Seems unlikely, considering LoRa.h does not support SX126x. So either your device is not SX1268 or you are not using LoRa.h

Your code is also rather suspicious. Here for example you are attempting to configure the device before it has been initialized:

```c++
  LoRa.setCurrentLimit(230.0);
  LoRa.setOutputPower(20);

  int state = LoRa.begin(433.775, 125.0 , 12 , 5 , 0x12, 20 , 20, 0 );
```

Had you not ignored the return codes of (nearly) all configuration methods, you would have discovered that some of them are failing.

Another strange thing is here - 0 is not an alias for "unused pin", it's a valid GPIO number. If some of your pins are note connected, you have to use `RADIOLIB_NC`. Keep in mind the SX127x cannot work without DIO0.

```
#ifdef RADIOLIB
  SX1278 LoRa = new Module(LORA_CS, 0 , LORA_RST,0);
#endif
```

The primary issue is that you have left the overcurrent protection level at the default value. But since you have made several mistakes in usage, I will recommend you go to the (quite extensive, IMHO) example collection here and use those as the starting point. Properly check return codes of all RadioLib methods, they are there to tell you exactly what went wrong.

**Reply by micfuc** [reply]
Hi,
thank you very much for your extensive comments... 
I just made a quick and DIRTY hack to check the power issue... by copying some examples from the examples directory  of the libraries... in my code for my designs I ALWAYS check for return codes... 
 
For the sx1268 it is unfortunatelly a writing error... it is  sx1278 clearly if you see the 20dbm special case that is a unique for the various lora chips around......

I will follow carefully your suggestions and just in case, if you like, I can report on my findings.

Best regards
Michele

**Reply by micfuc** [reply]
Hi Jan,
I must admit that I was too quick in my code hacking  :) ...  
The mistake was infact in coping the power and OCP lines BEFORE the module init... unfortunatelly ...
In the mean time I made my hack a little better and now it works in interrupt mode so that I have the possibility to see the power value on a small power meter very easily...
Now the power values for the old LoRa lib and RadioLib is exactly the same value ( obviously...).

In realty I have nearly never used sx127x chips till now but nearly always in my designs I use second generation chips sx126x... I was just now getting to the sx127x chip because I want to try seeing what can be done on 144 MHz by using some special HW I made for trying to work at very small BW   to stay in the regulatory limits we have ... then I was trying to see what power level the sx1278 was able to provide on 144MHz... this is why I was doing these tests... 
BTW a thing I found on the 144 MHz case is that the maximum power seems to be around 13 dbm ( with both the old and new LoRa libraries :) .. ) ...  
What do you thing about this power value ? 

BTW I thank you very much for your attention and quick answer... and for helping me to stay less quick in my hacks :)

I appreciate very much your  wonderfull RadioLib ... 
I am using it from several years in my circuits ... ( https://sarimesh.net )
Best regards
Michele I8FUC

**Reply by jgromes** [reply]
Glad I could help ;)

Regarding the output power at 144 MHz, I don't think there's anything in the SX1278 IC itself that could cause this (the datasheet claims that the circuit can go as low as 137 MHz). However, if I'm not mistaken, you are not using just the IC - you are using the T-Beam, which was most likely not designed for this frequency range. So it is very likely the antenna, the matching network, any possible RF switches etc are not tuned to this frequency, reflecting some of that power back into the SX1278. If you want to use it in a custom frequency band, you will most likely have to design your own board.

**Reply by micfuc** [reply]
Hi Jan,
thanks for fast feedback... I do agree with your analisys... actually my target is not to use necessarily the T-Beam but a custom HW design I made to test different LoRa modules... in any case my SW  is also targetted to run on the classical TTGO so I was starting from that case as the most reliable compared to my custom makeups... 

The actual idea to make my test is to use the ebyte E32-400M30S module that includes internally an sx1278 and is able to work paired with a similar module based on SX1268  ( the E22-400M30S ) .. I  did not jet make any test of this setup because I am waiting samples of the E32 from cina... 
In any case till now I have never  changed the LoRa module provided radio interface just to replicate the original manifacturer guides.

Thanks again for your help
Best regards
Michele
